src/Bicep.Core/Semantics/Namespaces/SystemNamespaceType.cs

```diff
@@ -296,7 +307,16 @@ private static IEnumerable<FunctionOverload> GetSystemOverloads(IFeatureProvider
 
             yield return new FunctionOverloadBuilder("last")
                 // TODO even with non-literal types, some type arithmetic could be performed
-                .WithReturnResultBuilder(TryDeriveLiteralReturnType("last", LanguageConstants.Any), LanguageConstants.Any)
+                .WithReturnResultBuilder((binder, fileResolver, diagnostics, arguments, argumentTypes) =>
+                {
+                    return argumentTypes[0] switch
+                    {
+                        TypedArrayType typedArrayType => new(typedArrayType.Item.Type),
+                        TupleType tupleType when tupleType.Items.Length == 0 => new(LanguageConstants.Null),
+                        _ => TryDeriveLiteralReturnType("last", LanguageConstants.Any)
+                                .Invoke(binder, fileResolver, diagnostics, arguments, argumentTypes)
+                    };
+                }, TypeHelper.CreateTypeUnion(LanguageConstants.Object, LanguageConstants.Any, LanguageConstants.Null))
```

Comment:
See comments on `first`. The return result builder for these two functions should be identical except for how non-empty tuples are handled (e.g., use `LastOrDefault` instead of `FirstOrDefault`).